### ISSUE TITLE
Startup race condition with GlusterFS.

### ISSUE THREAD

**Issue by phunyguy** [nfs-ganesha/nfs-ganesha#694]
I have run into a problem while setting up a single-node proof of concept for a hyperconverged Proxmox/Ganesha/GlusterFS setup that will eventually be 3 nodes.  Booting this server, GlusterFS and nfs-ganesha start in parallel, and if GlusterFS is not 100% initialized with the target volume just yet, nfs-ganesha fails with the following error:

```
10/03/2021 17:03:50 : epoch 60494246 : proxmox1 : ganesha.nfsd-2191[main] main :MAIN :EVENT :ganesha.nfsd Starting: Ganesha Version 2.7.1
10/03/2021 17:03:50 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] nfs_set_param_from_conf :NFS STARTUP :EVENT :Configuration file successfully parsed
10/03/2021 17:03:50 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] init_server_pkgs :NFS STARTUP :EVENT :Initializing ID Mapper.
10/03/2021 17:03:50 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] init_server_pkgs :NFS STARTUP :EVENT :ID Mapper successfully initialized.
10/03/2021 17:03:50 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] nfs_start_grace :STATE :EVENT :NFS Server Now IN GRACE, duration 90
10/03/2021 17:03:50 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] glusterfs_create_export :FSAL :EVENT :Volume temp exported at : '/'
10/03/2021 17:03:52 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] glusterfs_get_fs :FSAL :CRIT :Unable to initialize volume. Volume: temp
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] mdcache_fsal_create_export :FSAL :MAJ :Failed to call create_export on underlying FSAL GLUSTER
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] fsal_cfg_commit :CONFIG :CRIT :Could not create export for (/temp) to (/temp)
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] main :NFS STARTUP :WARN :No export entries found in configuration file !!!
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:33): 1 validation errors in block FSAL
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:33): Errors processing block (FSAL)
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:38): Unknown parameter (Up_poll_usec)
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:39): Unknown parameter (Transport)
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:11): 2 errors while processing parameters for EXPORT
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:11): Errors processing block (EXPORT)
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] lower_my_caps :NFS STARTUP :EVENT :CAP_SYS_RESOURCE was successfully removed for proper quota management in FSAL
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] lower_my_caps :NFS STARTUP :EVENT :currenty set capabilities are: = cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,cap_audit_read+ep
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] nfs_Init_svc :DISP :CRIT :Cannot acquire credentials for principal nfs
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] nfs_Init_admin_thread :NFS CB :EVENT :Admin thread initialized
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] nfs_rpc_cb_init_ccache :NFS STARTUP :WARN :gssd_refresh_krb5_machine_credential failed (-1765328160:22)
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] nfs_Start_threads :THREAD :EVENT :Starting delayed executor.
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] nfs_Start_threads :THREAD :EVENT :9P/TCP dispatcher thread was started successfully
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] nfs_Start_threads :THREAD :EVENT :gsh_dbusthread was started successfully
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[_9p_disp] _9p_dispatcher_thread :9P DISP :EVENT :9P dispatcher started
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] nfs_Start_threads :THREAD :EVENT :admin thread was started successfully
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] nfs_Start_threads :THREAD :EVENT :reaper thread was started successfully
10/03/2021 17:03:53 : epoch 60494246 : proxmox1 : ganesha.nfsd-2192[main] nfs_Start_threads :THREAD :EVENT :General fridge was started successfully
```
After discussing with boulos on IRC, they requested I submit this bug report to help facilitate modification to the GlusterFS FSAL, that will have it perform some kind of looping/checking to see if the glusterfs volume is available first, and just hang out and retry until it is, up to a timeout.  

EDIT:
I'll provide a little more clarification on the use case.  Single-node, or 3-node, hyperconverged or split, if the glusterfs volume is not available before nfs-ganesha starts up, this problem will occur.  If I were to create a dependency in systemd to start one before the other, it would work sometimes, but not every time, as there is sometimes a (very) slight delay in volume activation after glusterd starts, and nfs-ganesha receives the same result.  If they are on different servers, that result/failure could be even worse. If possible, there really should be a mechanism in place to monitor the requested glusterfs volume to see if it is up or down, and if down, wait until it is up, including if the volume went down *after* everything was working, in other words, reestablish.  The slightest hiccup otherwise causes problems, and requires a full nfs-ganesha service restart to fix.

Thanks for listening.

**Comment by ffilz** [comment]
@kalebskeithley could you take a look at this one?

**Comment by kalebskeithley** [comment]
You can add glusterd.service to the Wants= and After= lines in the nfs-ganesha.service file.

**Comment by phunyguy** [comment]
I've considered/tried that, and it still causes problems as there can be a slight delay in actual volume startup, so nfs-ganesha still gets an error if the volume isn't available yet.  It makes it a little better, as in it doesn't happen every time, but there are still random problems on reboot.  Also this doesn't work at all if nfs-ganesha and glusterfs are on different servers.

**Comment by ffilz** [comment]
Did we make any progress on this? We will be releasing Ganesha V4.0 soon with rc1 this week. If there is something we can address here it would be good to get it in.

**Comment by ffilz** [comment]
I'm going to mark this as an enhancement, anyone is free to work on this and submit patches but it is unlikely we will consider it for work.

**Comment by ffilz** [comment]
I don't think there is any resource to address this. I'm closing as wontfix. It can always be re-opened if someone finds reason to work on it.
